[rmlui] Add Rmlui library

mikke89 opened the pull request:
**Describe the pull request**

- #### What does your PR fix?  
  Adds a new port: [RmlUi](https://github.com/mikke89/RmlUi)

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
  Hopefully all platforms supported, tested on x64-windows, x64-windows-static, x86-windows.

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
  Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
  Yes

mikke89 committed:
Add RmlUi port

mikke89 committed:
Update versions

mikke89 committed:
Fix uwp build

mikke89 committed:
Update versions

JackBoosY commented on ports/rmlui/vcpkg.json:
```diff
@@ -0,0 +1,25 @@
+  "version": "4.0",
+  "description": "RmlUi is the C++ user interface library based on the HTML and CSS standards, designed as a complete solution for any project's interface needs.",
+  "homepage": "https://github.com/mikke89/RmlUi",
+  "documentation": "https://mikke89.github.io/RmlUiDoc/",
```
Since you are the owner, can you please add `maintainers` here? See docs:
https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#maintainers

mikke89 committed:
Add maintainers to manifest

mikke89 committed:
Update version

mikke89 commented:
I noticed some issues when consuming the library with static builds. I'm fixing things upstream so that the necessary define `RMLUI_STATIC_LIB` is properly exported in CMake.

My (unpublished) changes seem to work well for CMake builds now. However, if I may ask, is it correct that MSBuild does not make use of the exported defines from CMake, so they need to be added manually by users? What is the best course of action in vcpkg ports to make sure macros are included, should we patch the library source to add the macro definition in-source?

Please don't merge before this is resolved, thanks!

JackBoosY commented:
> My (unpublished) changes seem to work well for CMake builds now. However, if I may ask, is it correct that MSBuild does not make use of the exported defines from CMake, so they need to be added manually by users? What is the best course of action in vcpkg ports to make sure macros are included, should we patch the library source to add the macro definition in-source?

If the macro exists in the header file, you should use `vcpkg_replace_string` to replace this macro in the header file with `#if 1` according to the conditions. See example:
https://github.com/microsoft/vcpkg/blob/030cfaa24de9ea1bbf0a4d9c615ce7312ba77af1/ports/libflac/portfile.cmake#L56-L74

noam-c commented on ports/rmlui/portfile.cmake:
```diff
@@ -0,0 +1,37 @@
+        ${FEATURE_OPTIONS}
+)
+
+vcpkg_install_cmake()
+vcpkg_fixup_cmake_targets(
+    CONFIG_PATH    lib/RmlUi/cmake
+)
+vcpkg_copy_pdbs()
```
@mikke89 This function has been deprecated, unfortunately.
Luckily I've been spending some time this week _also_ trying to create a port for RmlUi. :joy:

Here's what worked for me:
```suggestion
vcpkg_cmake_install()
vcpkg_cmake_config_fixup(
    CONFIG_PATH lib/${PORT}/cmake
)

vcpkg_copy_pdbs()
```

noam-c replied:
(note for the new commands I needed to add `vcpkg-cmake-config` and `vcpkg-cmake` as dependencies in the `vcpkg.json`)

mikke89 replied:
Cool, thanks! I'll make sure to get those changes in.

mikke89 committed:
Update ports/rmlui/portfile.cmake

Co-authored-by: Noam Chitayat <noam-c@users.noreply.github.com>

mikke89 committed:
- Update to RmlUi 4.1, fixes some issues with missing dependencies.
- Add robin-hood-hashing as dependency, replacing built-in copy.
- Only need FreeType core.
- Static library definition now replaced in-source.

mikke89 committed:
Fix config name

JackBoosY commented on ports/rmlui/portfile.cmake:
```diff
@@ -0,0 +1,57 @@
+		freetype        NO_FONT_INTERFACE_DEFAULT
+)
+
+if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
+	vcpkg_replace_string(${SOURCE_PATH}/Include/RmlUi/Core/Header.h
+		"#if !defined RMLUI_STATIC_LIB"
+		"#if 0"
+	)
+	vcpkg_replace_string(${SOURCE_PATH}/Include/RmlUi/Debugger/Header.h
+		"#if !defined RMLUI_STATIC_LIB"
+		"#if 0"
+	)
+	if ("lua" IN_LIST FEATURES)
+		vcpkg_replace_string(${SOURCE_PATH}/Include/RmlUi/Lua/Header.h
+			"#if !defined RMLUI_STATIC_LIB"
+			"#if 0"
+		)
+	endif()
+endif()
```
I think this should be moved to `vcpkg_cmake_install` and modify the content in _${CURRENT_PACKAGES_DIR}/include_. Modifications here will affect the entire build.

mikke89 replied:
You're right, thanks. I've changed it now.

JackBoosY commented on ports/rmlui/portfile.cmake:
```diff
@@ -0,0 +1,57 @@
+
+file(REMOVE_RECURSE 
+	${CURRENT_PACKAGES_DIR}/debug/include
+	${CURRENT_PACKAGES_DIR}/debug/lib/RmlUi
+	${CURRENT_PACKAGES_DIR}/lib/RmlUi
```
Is this a header-only library? The _lib_ folder should exist when building static.

mikke89 replied:
This just removes some empty directories which originally contained the CMake target files. The library files are still there with the names `RmlCore` and `RmlDebugger`.

mikke89 committed:
Change source in current packages directory

JackBoosY commented on ports/rmlui/portfile.cmake:
```diff
@@ -0,0 +1,57 @@
+)
+
+# Replace built-in thirdparty header
+file(COPY
+	${CURRENT_INSTALLED_DIR}/include/robin_hood.h
+	DESTINATION ${SOURCE_PATH}/Include/RmlUi/Core/Containers
```
Which port provide this header? I think we should use `find_path` in `CMakeLists.txt`, remove the old one and add this file path to `target_include_directories` instead.

mikke89 replied:
Alright, let me know if I understood correctly with the changes I made. It required a bit of patch-work, but if you prefer it this way then it's all good to me.

mikke89 committed:
Use robin_hood.h included from vcpkg

JackBoosY commented on ports/rmlui/portfile.cmake:
```diff
@@ -0,0 +1,64 @@
+		freetype        NO_FONT_INTERFACE_DEFAULT
+)
+
+# Replace built-in header with vcpkg version (from robin-hood-hashing port)
+file(REMOVE ${SOURCE_PATH}/Include/RmlUi/Core/Containers/robin_hood.h)
+vcpkg_replace_string( ${SOURCE_PATH}/Include/RmlUi/Config/Config.h
+	"#include \"../Core/Containers/robin_hood.h\""
+	"#include <robin_hood.h>"
+)
+vcpkg_replace_string( ${SOURCE_PATH}/CMake/FileList.cmake
+	"\${PROJECT_SOURCE_DIR}/Include/RmlUi/Core/Containers/robin_hood.h"
+	""
+)
```
Can you please also patch these changes into `add-robin-hood.patch`?

mikke89 committed:
Move source modifications to robin_hood patch

JackBoosY commented:
When building rmlui:x64-linux:
```
CMake Error at scripts/cmake/z_vcpkg_apply_patches.cmake:61 (message):
  Applying patch failed: Checking patch CMakeLists.txt...

  Checking patch cmake/FileList.cmake...

  error: cmake/FileList.cmake: No such file or directory

  Checking patch include/RmlUi/Config/Config.h...

  error: include/RmlUi/Config/Config.h: No such file or directory

Call Stack (most recent call first):
  scripts/cmake/vcpkg_extract_source_archive_ex.cmake:144 (z_vcpkg_apply_patches)
  scripts/cmake/vcpkg_from_github.cmake:154 (vcpkg_extract_source_archive_ex)
  ports/rmlui/portfile.cmake:1 (vcpkg_from_github)
  scripts/ports.cmake:139 (include)
```

mikke89 committed:
Fix path in patch

ras0219-msft commented:
This LGreatTM, thanks for the awesome library and making a PR into vcpkg!

mikke89 commented:
Thank you!
